metaflow/plugins/tagging_cli.py

```diff
@@ -0,0 +1,424 @@
+        # to make it consistent with the by_run listing
+        obj.echo_always(' ; '.join([
+            'system:%s' % cat if is_system else 'user:%s' % cat,
+            ','.join(runs)]))
+
+
+def print_tags(obj, metaflow_run_client):
+    system_tags = {metaflow_run_client.pathspec: metaflow_run_client.system_tags}
+    all_tags = {metaflow_run_client.pathspec: metaflow_run_client.tags}
+    return format_tags(obj, system_tags, all_tags, by_tag=False)
+
+
+def is_prod_token(token):
+    return token is not None and token.startswith("production:")
+
+
+def obtain_metaflow_run_client(obj, run_id, user_namespace):
+    if is_prod_token(user_namespace):
+        raise CommandException(
+            "Modifying a scheduled run's tags is currently not allowed.")
+
+    flow_name = obj.flow.name
+
+    # handle error messaging for two cases
+    # 1. our user tries to tagging a new flow before it is run
+    # 2. our user makes a typo in --namespace
+    try:
+        namespace(user_namespace)
+        Flow(pathspec=flow_name)
+    except MetaflowNotFound:
+        raise CommandException("Flow *%s* does not exist " % flow_name +
+                               "in any namespace.\n" +
+                               "Please run the flow before tagging.")
+    except MetaflowNamespaceMismatch:
+        raise CommandException("Flow *%s* does not belong to " % flow_name +
+                               "namespace *%s*.\n" % user_namespace +
+                               "Check typos if you used option --namespace.")
+
+    # throw an error with message to include latest run-id when run_id is None
+    if run_id is None:
+        latest_run_id = Flow(pathspec=flow_name).latest_run.id
+        msg = "Please specify a run-id using --run-id\n" + \
+              "Flow %s's latest run has run-id *%s* under namespace %s" \
+            % (flow_name, latest_run_id, user_namespace)
+        raise CommandException(msg)
+    run_id_parts = run_id.split('/')
+    if len(run_id_parts) == 1:
+        run_name = "%s/%s" % (flow_name, run_id)
+    elif len(run_id_parts) == 2:
+        if run_id_parts[0] != flow_name:
+            raise CommandException(
+                "Requesting run-id *%s* which is not a run of flow %s" % (run_id, flow_name))
+        run_name = run_id
+    else:
+        raise CommandException("Run-id *%s* is not a valid run-id" % run_id)
+
+    # handle error messaging for three cases
+    # 1. our user makes a typo in --run-id
+    # 2. our user tries to mutate tags on a meson/production run
+    # 3. our user's --run-id does not exist in the default/specified namespace
+    try:
+        namespace(user_namespace)
+        metaflow_client = Run(pathspec=run_name)
+    except MetaflowNotFound:
+        raise CommandException("Run *%s* does not exist " % run_name +
+                               "in any namespace.\nPlease check your --run-id.")
+    except MetaflowNamespaceMismatch:
+        namespace(None)
```

Comment:
What's the intention here?

Reply:
Simply to generate a proper error message with useful info. There's `MetaflowNamespaceMismatch` exception when the `run id` user specified does not live in the namespace. We will then try to locate the correct namespace and suggest to user in the error message